File: Algoritm2_C/sort.c
#include <stdio.h>
#include <stdlib.h>
#include "sort.h"

void interchange_values(int *address_1, int *address_2)
{
    int aux;
    aux = *address_1;
    *address_1 = *address_2;
    *address_2 = aux;
}

int partition(int *vector, int left, int right)
{
    int pivot = vector[left];
    int i = left - 1;
    int j = right + 1;

    while (1)
    {
        do
        {
            i++;
        }
        while (*(vector + i) < pivot);

        do
        {
            j--;
        }
        while (*(vector + j) > pivot);

        if (i >= j)
            return j;

        interchange_values(vector + i, vector + j);
    }
}

void quick_sort(int *vector, int left, int right)
{
    int pivot_position = 0;

    if (left < right)
    {
        pivot_position = partition(vector, left, right);
        quick_sort(vector, left, pivot_position);
        quick_sort(vector, pivot_position + 1, right);
    }
}

File: Algoritm2_Py/Algoritm2_Python/sort.py
def partition(vector, left, right):
    i = (left - 1)
    pivot = vector[right]

    for j in range(left, right):
        if vector[j] <= pivot:
            i = i + 1
            vector[i], vector[j] = vector[j], vector[i]

    vector[i + 1], vector[right] = vector[right], vector[i + 1]
    return (i + 1)


def quick_sort(vector,  left, right):
    if left < right:
        pi = partition(vector, left, right)
        quick_sort(vector, left, pi - 1)
        quick_sort(vector, pi + 1, right)

